src/components/input/input.ts

```diff
@@ -0,0 +1,108 @@
+/**
+ * Grow Elastic TextArea
+ *
+ * Increases the `rows` attribute of a `textarea` until it no longer scrolls.
+ *
+ * @param {HTMLTextAreaElement} textarea - the target `textarea` element
+ */
+const growElasticTextArea = (textarea: HTMLTextAreaElement) => {
+  const minRows = Number(textarea.dataset.minRows);
+  let rows = Number(textarea.getAttribute('rows')) || minRows;
+  let isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+  // Grow until we stop scrolling
+  while (isScrolling) {
+    rows += 1;
+    textarea.setAttribute('rows', String(rows));
+    isScrolling = textarea.scrollHeight > textarea.clientHeight;
+  }
+};
+
+/**
+ * Shrink Elastic TextArea
+ *
+ * Decreases the `rows` attribute of a `textarea` until it begins scrolling.
+ *
+ * @param {HTMLTextAreaElement} textarea - the target `textarea` element
+ */
+const shrinkElasticTextArea = (textarea: HTMLTextAreaElement) => {
+  const minRows = Number(textarea.dataset.minRows);
+  let rows = Number(textarea.getAttribute('rows')) || minRows;
+  let isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+  // Shrink until we hit the minimum rows or start scrolling
+  for (let i = rows; i >= minRows; i--) {
+    rows = i;
+    textarea.setAttribute('rows', String(Math.max(rows, minRows)));
+    isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+    // If we're scrolling, then we shrank too much, so grow a bit and stop
+    if (isScrolling) {
+      growElasticTextArea(textarea);
+      break;
+    }
+  }
+};
+
+/**
+ * Update Elastic TextArea
+ *
+ * When a textarea is updated, compares the new value to the previous value
+ * (from a dataset attribute) to determine whether to grow or shrink.
+ *
+ * Note: Ideally we'd destructure `event.target` from the parameters and declare
+ * the type directly, since we don't care about the rest of the `event` object.
+ * Unfortunately, that caused some TS problems. If someone finds a solution,
+ * we should update this function to take `{target}` as the only parameter.
+ * @see https://stackoverflow.com/a/28900856/4898045
+ *
+ * @param {Event} event - the input event on the target `textarea`
+ */
+const updateElasticTextArea = (event: Event) => {
+  const textarea = event.target as HTMLTextAreaElement;
+  const lastValue = textarea.dataset.lastValue || '';
```

Comment:
Instead of storing the last value in a data attribute, I'm wondering if we could store it as an "instance variable" within the `createElasticTextArea` instance and move `update` into the `createElasticTextArea` function:

```js
export const createElasticTextArea = () => {
  let lastLength = textarea.value.length
  const update = (event) => {
    if (lastLength > ...) {}
    // ...
    lastLength = element.length
  }
  // ... listeners
}
```

I think it might also make sense to move `grow` and `shrink` into the `createElasticTextArea` function, I'd be curious to hear your thoughts

Reply:
> Instead of storing the last value in a data attribute, I'm wondering if we could store it as an "instance variable" within the createElasticTextArea instance and move update into the createElasticTextArea function:

I agree with this. 👍

Reply:
> I think it might also make sense to move grow and shrink into the createElasticTextArea function, I'd be curious to hear your thoughts

I was wondering about this as well. It feels like they are a part of the same component. If the `grow` and/or `shrink` functions were reusable elsewhere, then they might live as their own functions and possibly be exported from their own module. But since they are only related to the `createElasticTextArea`, I think it'll be most clear for them to _literally_ live within the `createElasticTextArea` scope. 😉

Reply:
I had the same thought. Then we can define all of our variables once and avoid needing to mess with the data attributes in the DOM.

Reply:
Sounds good. I was also unhappy with having to repeat all the variables in grow and shrink, so this approach would solve that. I'll refactor.